crates/feldera-storage/src/file/writer.rs

```diff
@@ -0,0 +1,1031 @@
+            n_values: n_values as u32,
+            value_map_varint,
+            row_group_varint,
+            value_map_ofs: specs.value_map.start as u32,
+            row_groups_ofs,
+        };
+        header.overwrite_head(&mut self.raw);
+
+        DataBlock {
+            raw: self.raw,
+            min_offset: *self.value_offsets.first().unwrap(),
+            max_offset: *self.value_offsets.last().unwrap(),
+            n_values,
+        }
+    }
+}
+
+struct IndexEntry {
+    child: BlockLocation,
+    min_offset: usize,
+    max_offset: usize,
+    row_total: u64,
+}
+
+struct ContiguousRanges(Vec<u64>);
+
+impl ContiguousRanges {
+    fn with_capacity(capacity: usize) -> Self {
+        Self(Vec::with_capacity(capacity.saturating_add(1)))
+    }
+    fn push(&mut self, range: &Range<u64>) {
+        match self.0.last() {
+            Some(&last) => {
+                debug_assert_eq!(last, range.start);
+            }
+            None => self.0.push(range.start),
+        };
+        self.0.push(range.end);
+    }
+    fn pop(&mut self) {
+        match self.0.len() {
+            0 | 1 => unreachable!(),
+            2 => self.0.clear(),
+            _ => {
+                self.0.pop();
+            }
+        }
+    }
+    fn max(&self) -> Option<u64> {
+        self.0.last().copied()
+    }
+}
+
+struct IndexBlockBuilder {
+    parameters: Rc<Parameters>,
+    column_index: usize,
+    raw: AlignedVec,
+    entries: Vec<IndexEntry>,
+    child_type: NodeType,
+    size_target: Option<usize>,
+}
+
+struct IndexBuildSpecs {
+    bound_map: VarintWriter,
+    row_totals: VarintWriter,
+    child_pointers: VarintWriter,
+    len: usize,
+}
+
+struct IndexBlock {
+    raw: AlignedVec,
+    min_offset: usize,
+    max_offset: usize,
+    n_rows: u64,
+}
+
+fn rkyv_copy_key<K, A>(dst: &mut AlignedVec, src: &AlignedVec, offset: usize) -> usize
+where
+    K: Rkyv,
+    A: Rkyv,
+{
+    let archived: &Archived<Item<K, A>> =
+        unsafe { archived_value::<Item<K, A>>(src.as_slice(), offset) };
+    let value: K = archived.0.deserialize(&mut Infallible).unwrap();
+    rkyv_serialize(dst, &value)
+}
+
+fn rkyv_copy<T>(dst: &mut AlignedVec, src: &AlignedVec, offset: usize) -> usize
+where
+    T: Rkyv,
+{
+    let archived: &T::Archived = unsafe { archived_value::<T>(src, offset) };
+    let value: T = archived.deserialize(&mut Infallible).unwrap();
+    rkyv_serialize(dst, &value)
+}
+
+fn rkyv_serialize<T>(dst: &mut AlignedVec, value: &T) -> usize
+where
+    T: Archive + for<'a> Serialize<Serializer<'a>>,
+{
+    let old_len = dst.len();
+
+    let mut serializer = Serializer::new(
+        AlignedSerializer::new(take(dst)),
+        Default::default(),
+        Default::default(),
+    );
+    let offset = serializer.serialize_value(value).unwrap();
+    *dst = serializer.into_components().0.into_inner();
+
+    if dst.len() == old_len {
+        // Ensure that a value takes up at least one byte.  Otherwise, we'll
```

Comment:
does this warrant some warning?

Reply:
I think it's likely only to come up if the key type is `()`, or for some similarly empty type. I don't know whether it's worth it. It might be too much paranoia even checking.